localstack/aws/handlers/service.py

```diff
@@ -218,6 +220,71 @@ def create_exception_response(self, exception: Exception, context: RequestContex
             status_code = 501 if config.FAIL_FAST else 500
 
             error = CommonServiceException("InternalError", msg, status_code=status_code)
+            context.service_exception = error
 
         serializer = create_serializer(context.service)  # TODO: serializer cache
         return serializer.serialize_error_to_response(error, operation)
+
+
+class ServiceResponseParser(Handler):
+    """
+    This response handler makes sure that, if the current request in an AWS request, that either ``service_response``
+    or ``service_exception`` of ``RequestContext`` is set to something sensible before other downstream response
+    handlers are called. When the Skeleton invokes an ASF-native provider, this will mostly return immediately
+    because the skeleton sets the service response directly to what comes out of the provider. When responses come
+    back from backends like Moto, we may need to parse the raw HTTP response, since we sometimes proxy directly. If
+    the ``service_response`` is an error, then we parse the response and create an appropriate exception from the
+    error response. If ``service_exception`` is set, then we also try to make sure the exception attributes like
+    code, sender_fault, and message have values.
+    """
+
+    def __call__(self, chain: HandlerChain, context: RequestContext, response: Response):
+        if not context.operation:
+            return
+
+        if context.service_response:
+            return
+
+        if exception := context.service_exception:
+            if isinstance(exception, ServiceException):
+                try:
+                    exception.code
+                except AttributeError:
+                    # FIXME: we should set the exception attributes in the scaffold when we generate the exceptions.
+                    #  this is a workaround for now, since we are not doing that yet, and the attributes may be unset.
+                    self._set_exception_attributes(context.operation, exception)
```

Comment:
We will be able to remove this Tuesday latest (after https://github.com/localstack/localstack/pull/6409 is merged, and the result of our update action has been merged in all repos). 🥳